### ISSUE TITLE
Inconsistent Generation for BNB 4Bit Llama-7B on Transformers vs TGI

### ISSUE THREAD

**Issue by rib-2** [huggingface/text-generation-inference#1074]
### System Info

- OS Version: Ubuntu 22.04.2 LTS
- Hardware: 1 Nvidia A10 GPU on AWS g5.4xlarge instance
- Deployment specificities: Just developing on a VM
- Version: Running with official TGI Docker Image 1.1.0

Launch Command:
`docker run --gpus device=0 --shm-size 1g --network host -v $tgi_volume:/data -e HUGGING_FACE_HUB_TOKEN=$hf_token -it ghcr.io/huggingface/text-generation-inference:1.1.0 --model-id meta-llama/Llama-2-7b-hf --quantize bitsandbytes-nf4`

### Information

- [X] Docker
- [ ] The CLI directly

### Tasks

- [X] An officially supported command
- [ ] My own modifications

### Reproduction

I was attempting to verify that TGI generates the same responses as my local transformers model when quantizing with Bitsandbytes NF4.

Using transformers directly, I saw the following example generation:

- Transformers Version: `4.33.3`
- Bitsandbytes Version:  `0.41.1`

```python
import torch
from transformers import (
    AutoModelForCausalLM, 
    AutoTokenizer, 
    BitsAndBytesConfig,
    GenerationConfig
)

EXAMPLE_PROMPT = "In a galaxy far, far away,"
MODEL_ID = "meta-llama/Llama-2-7b-hf"

# load model
model_transformers = AutoModelForCausalLM.from_pretrained(
    MODEL_ID,
    load_in_4bit=True,
    torch_dtype=torch.float16,
    device_map="auto",
    quantization_config=BitsAndBytesConfig(
        load_in_4bit=True,
        bnb_4bit_compute_dtype=torch.float16,
        bnb_4bit_use_double_quant=True,
        bnb_4bit_quant_type='nf4',
    ),
)
tokenizer = AutoTokenizer.from_pretrained(MODEL_ID)

# tokenize
model_inputs = tokenizer(EXAMPLE_PROMPT, return_tensors="pt").to("cuda")
generation_config = GenerationConfig(
    max_new_tokens=100,
    do_sample=False
)

# generate text
output = model_transformers.generate(
    **model_inputs,
    generation_config=generation_config
)
print(tokenizer.batch_decode(output, skip_special_tokens=True)[0])
```

The output is as follows:
```bash
In a galaxy far, far away, a long time ago, a young boy named Luke Skywalker was born. nobody knows what happened to him after that.
The first time I saw Star Wars was when I was 10 years old. I was in the 5th grade. I was in the library, and I was reading a book. I was reading a book about the history of the Roman Empire. I was reading a book about the history of the Roman Empire. I was reading a book about the history of
```

However, when launching a text generation server with nf-4 quantization on the same model, I get a different generation result with the same greedy decoding strategy:

Launch server:
```bash
docker run --gpus all --shm-size 1g --network host -v $tgi_volume:/data -e HUGGING_FACE_HUB_TOKEN=$hf_token -it ghcr.io/huggingface/text-generation-inference:1.1.0 --model-id meta-llama/Llama-2-7b-hf --quantize bitsandbytes-nf4
```

Make request:
```bash
curl 127.0.0.1:80/generate \
    -X POST \
    -d '{"inputs":"In a galaxy far, far away,","parameters":{"max_new_tokens":100,"do_sample":false}}' \
    -H 'Content-Type: application/json'
```

The output is as follows:
```bash
{"generated_text":" the Empire is in control. everybody is under the control of the Empire. The only hope for the galaxy is the Rebel Alliance. The Rebels are fighting the Empire to free the galaxy from the control of the Empire. The Rebels are led by a man named Luke Skywalker. Luke is a young man who is trying to find his place in the galaxy. He is trying to find his place in the galaxy by joining the Rebel Alliance. The Rebels are"}
```

### Expected behavior

I would expect the generated text from the server with the same quantization scheme to match the generation form local transformers

**Comment by Narsil** [comment]
Have you tried removing 

```
 bnb_4bit_compute_dtype=torch.float16,
 bnb_4bit_use_double_quant=True,
 ```
 
 TGI uses all defaults (I'm not sure what `double_quant` stands for there).
 
 Also it's perfectly ok if they don't generate the same thing, given it's not the same codebase, therefore subtle differences in logits can happen (they often do). What's more suprising is that they happen right away, so there's definitely something more that could be at play here.

**Comment by rsnm2** [comment]
Yeah it makes sense there is some differences especially with the custom kernels and whatnot, I just think the magnitude of the difference was more than I expected

- The default dtype for the model in TGI is fp16, so I don't think that is the issue
- The double_quantization refers to quantizing the weights and scales. I believe TGI sets this argument to true as `compress_statistics=true` in the `Linear4Bit` constructor at https://github.com/huggingface/text-generation-inference/blob/main/server/text_generation_server/utils/layers.py#L236

I will dig in a bit more and compute some statistics about logits difference and look into some layer by layer items to see if I can isolate it

**Comment by Narsil** [comment]
@rsnm2 If you can log the input_ids in both versions that'd help too, I'm wondering if it couldn't be there already.

**Comment by rsnm2** [comment]
Quick update

I have narrowed the issue down to the attention layer

What I am seeing is:
- hidden state passed to the first layer is consistent for transformers and tgi
- hidden state passed to the second layer and beyond diverge w/ l2 norms >0.05 

Looking closer, I am seeing some numerical differences in the `input_layernorm` calculation. TGI uses flash attention's `dropout_layer_norm.dropout_add_ln_fwd` while transformers uses a torch implementation. I switched out the flash attention implmentation for the torch implementation, and this removed the numerical differences for the input layernorm

After doing this, the input to the query, key, value calculations are exactly the same across transformers and tgi, but I am seeing large divergence in the results between transformers and tgi

```python
# tgi model, note --- uses modified layernorm to match transformers implementation
layer = tgi_model.model.model.layers[0]
normed_tgi, _ = layer.input_layernorm(layer.hidden_states, layer.residual)

# transformers model
ts_layer = model_transformers.model.layers[0]
normed_ts = ts_layer.input_layernorm(output.hidden_states[0][0][0])

# confirm inputs are the same
print((normed_tgi - normed_ts).norm())
# >> tensor(0., device='cuda:0', dtype=torch.float16)

# tgi model q,k,v
qkv = layer.self_attn.query_key_value(normed_tgi)
query, key, value = qkv.split(
    [
        layer.self_attn.head_size * layer.self_attn.num_heads,
        layer.self_attn.head_size * layer.self_attn.num_key_value_heads,
        layer.self_attn.head_size * layer.self_attn.num_key_value_heads,
    ],
    dim=1,
)

# transformers model q,k,v
query_ts = ts_layer.self_attn.q_proj(normed_ts)
key_ts = ts_layer.self_attn.k_proj(normed_ts)
value_ts = ts_layer.self_attn.v_proj(normed_ts)

# compute differences
print((query_ts - query).norm())
# >> tensor(0.6914, device='cuda:0', dtype=torch.float16)

print((key_ts - key).norm())
# >> tensor(0.6172, device='cuda:0', dtype=torch.float16)

print((value_ts - value).norm())
# >> tensor(0.0218, device='cuda:0', dtype=torch.float16)
```

Looking into how the `layer.self_attn.query_key_value` works in TGI, I see that there is one large weight matrix representing Wq, Wk, and Wv. As such, when we apply the quantization, under the hood there is one large `Params4Bit` weight created (https://github.com/huggingface/text-generation-inference/blob/main/server/text_generation_server/utils/layers.py#L233). As a result, in TGI, all of Wq, Wk, and Wv are quantized as one large tensor rather Wq, Wk, and Wv being quantized as three individual tensors (as they are in transformers). 

The exact details of the bitsandbytes quantization scheme are a bit past my current understanding, but I am concerned quantizing the weight matrices jointly may result in different values than quantizing separately, as 4bit quantization typically relies on grouping weights into blocks with scales that are determined by the weights in those blocks. It is possible that we have some blocks created which overlap between the separate matrices

This is just a hypothesis and I am not sure if this is the issue, but I need to jump back to my day job, so I will investigate over the weekend

**Comment by rsnm2** [comment]
I have found the source of the discrepancy between transformers and TGI for 4bit quantized models

**TLDR** 
The ultimate cause of the issue is that Torch does not guarantee that (A@B)[0] (the first element of the batched result) is bitwise identical to A[0]@B[0] (the matrix product of the first elements of the input batches) even though mathematically it’s an identical computation. 

This creates issues because in TGI the query, key, value states are computed in one large linear layer while in Transformers they are computed as three separate linear.

**Longer Investigation:**

- I found that the inputs to the first layer for TGI and Transformers is consistent, meaning there is no issue with tokenization or embeddings.

- I found some numerical differences in input_layernorm calculation for TGI vs transformers. TGI uses flash_attentions optimized version of RMSNorm while Transformers uses a Torch implementation. I resolved this difference by swapping TGI's implementation for Transformers. However, after doing this, the logit differences remained.

- With the input_layernorm discrepancies fixed, the inputs to the attention calculation are now exactly the same. I then found some large differences in the query, key, value states

```bash
layer = tgi_model.model.model.layers[0]
ts_layer = transformers_model.model.layers[0]

normed_tgi, _ = layer.input_layernorm(layer.hidden_states, layer.residual)
normed_ts = ts_layer.input_layernorm(output_transformers.hidden_states[0][0][0])

normed_ts = ts_layer.input_layernorm(output_transformers.hidden_states[0][0][0])

print("Input to Q,K,V Calculation Values Are The Same:")
print((normed_tgi - normed_ts).norm())
# >>> tensor(0., device='cuda:0', dtype=torch.float16)

qkv = layer.self_attn.query_key_value(normed_tgi)
query, key, value = qkv.split(
    [
        layer.self_attn.head_size * layer.self_attn.num_heads,
        layer.self_attn.head_size * layer.self_attn.num_key_value_heads,
        layer.self_attn.head_size * layer.self_attn.num_key_value_heads,
    ],
    dim=1,
)

print("\nQuery Key Value States Different:")
query_ts = ts_layer.self_attn.q_proj(normed_ts)
key_ts = ts_layer.self_attn.k_proj(normed_ts)
value_ts = ts_layer.self_attn.v_proj(normed_ts)

print((query_ts - query).norm())
print((key_ts - key).norm())
print((value_ts - value).norm())

# >>> tensor(0.0389, device='cuda:0', dtype=torch.float16)
# >>> tensor(0.0310, device='cuda:0', dtype=torch.float16)
# >>> tensor(0.0020, device='cuda:0', dtype=torch.float16)
```

Trying to debug this, I first looked into whether there were any issues actually quantizing the weights as one large matrix rather than as three separate matrices. However, looking into more detail at bnb's quantization scheme, I found that the default is to quantize in groups of 64 --> which meant that the quantized values were the exact same across the transformers and TGI implementations. There could be some small differences in cases where double quantization is turned on (which means that the metadata is also quantized) - this is on in TGI - but I turned this off. This meant that there is some numerical difference in the actual 4bit matrix multiplication:

```python
import bitsandbytes as bnb

linear_qkv_tgi = tgi_model.model.model.layers[0].self_attn.query_key_value.linear
linear_q_ts = transformers_model.model.layers[0].self_attn.q_proj
linear_k_ts = transformers_model.model.layers[0].self_attn.k_proj
linear_v_ts = transformers_model.model.layers[0].self_attn.v_proj

cutoffs = linear_qkv_tgi.weight.shape[0] // 3
print("All weights are the same:")
print((linear_qkv_tgi.weight[:cutoffs] == linear_q_ts.weight).all())
print((linear_qkv_tgi.weight[cutoffs:cutoffs*2] == linear_k_ts.weight).all())
print((linear_qkv_tgi.weight[cutoffs*2:cutoffs*3] == linear_v_ts.weight).all())

x = normed_tgi
out_tgi_qkv = bnb.matmul_4bit(
    x, linear_qkv_tgi.weight.t(), bias=None, quant_state=linear_qkv_tgi.weight.quant_state
)

out_ts_q = bnb.matmul_4bit(
    x, linear_q_ts.weight.t(), bias=None, quant_state=linear_q_ts.weight.quant_state
)

out_ts_k = bnb.matmul_4bit(
    x, linear_k_ts.weight.t(), bias=None, quant_state=linear_k_ts.weight.quant_state
)

out_ts_v = bnb.matmul_4bit(
    x, linear_v_ts.weight.t(), bias=None, quant_state=linear_v_ts.weight.quant_state
)

print("\nBut delta in the outputs")
print((out_tgi_qkv[:,4096*0:4096*1] - out_ts_q).norm())
print((out_tgi_qkv[:,4096*1:4096*2] - out_ts_k).norm())
print((out_tgi_qkv[:,4096*2:4096*3] - out_ts_v).norm())
```

Output:

```bash
All weights are the same:
tensor(True, device='cuda:0')
tensor(True, device='cuda:0')
tensor(True, device='cuda:0')

But delta in the outputs
tensor(0.0389, device='cuda:0', dtype=torch.float16)
tensor(0.0310, device='cuda:0', dtype=torch.float16)
tensor(0.0020, device='cuda:0', dtype=torch.float16)
```

So, I dug into the implementation of bnb.matmul4bit. Ultimately I traced it back to this ([line]https://github.com/TimDettmers/bitsandbytes/blob/main/bitsandbytes/autograd/_functions.py#L516)

```python
output = torch.nn.functional.linear(A, F.dequantize_4bit(B, state).to(A.dtype).t(), bias)

```

**Comment by Narsil** [comment]
Wow thanks a lot for this very deep dive, your findings are enlightening ! 

Given what you're saying I'm not sure we should do anything.
Do you think there's anything we should be doing ?
Fusing qkv, gives a significant performance boost at inference time, and there's no reason bnb should be that sensitive to it.

**Comment by rsnm2** [comment]
I think you are right --- especially since its not even bnb that is causing the issue (as the analysis above shows, the quant/dequant is working as expected)

Im still working through the full analysis though. 

Right now, after running the QKV at FP64 (just to get rid of the numerical issue during debugging), I'm still getting quite different logits. The divergence starts after RoPE is applied so I'm just going to try to see whats going on (any maybe even switch over to an A100)

Ill make a suggestion once I get through the full analysis

Thanks for listening to my ranting!

**Comment by Narsil** [comment]
Btw, since  you're diving that deep, you should still take a look at AWQ (or GTPQ) both of which have much faster runtimes than nf4.

For the RoPE, we do use the flash attention kernel implementation there, so there might be some other differences there.

**Comment by rsnm2** [comment]
Thanks for the tip!
